src/Tes.Runner/Events/EventsPublisher.cs

```diff
@@ -0,0 +1,225 @@
+            { "errorMessage", errorMessage??string.Empty}
+        };
+        await PublishAsync(eventMessage);
+    }
+
+    public virtual async Task PublishDownloadStartEventAsync(NodeTask nodeTask)
+    {
+        var eventMessage = CreateNewEventMessage(nodeTask.Id, DownloadStartEvent, StartedStatus,
+                       nodeTask.WorkflowId);
+
+        eventMessage.EventData = new Dictionary<string, string>();
+
+        await PublishAsync(eventMessage);
+    }
+
+    public virtual async Task PublishDownloadEndEventAsync(NodeTask nodeTask, int numberOfFiles, long totalSizeInBytes, string statusMessage, string? errorMessage = default)
+    {
+        var eventMessage = CreateNewEventMessage(nodeTask.Id, DownloadEndEvent, statusMessage,
+                       nodeTask.WorkflowId);
+        eventMessage.EventData = new Dictionary<string, string>
+        {
+            { "numberOfFiles", numberOfFiles.ToString()},
+            { "totalSizeInBytes", totalSizeInBytes.ToString()},
+            { "errorMessage", errorMessage??string.Empty}
+        };
+        await PublishAsync(eventMessage);
+    }
+
+    public async Task PublishTaskCompletionEventAsync(NodeTask tesNodeTask, TimeSpan duration, string statusMessage, string? errorMessage)
+    {
+        var eventMessage = CreateNewEventMessage(tesNodeTask.Id, TaskCompletionEvent, statusMessage,
+            tesNodeTask.WorkflowId);
+        eventMessage.EventData = new Dictionary<string, string>
+        {
+            { "duration", duration.ToString()},
+            { "errorMessage", errorMessage??string.Empty}
+        };
+
+        await PublishAsync(eventMessage);
+    }
+
+    private EventMessage CreateNewEventMessage(string? entityId, string name, string statusMessage,
+        string? correlationId)
+    {
+        return new EventMessage
+        {
+            Id = Guid.NewGuid().ToString(),
+            Name = name,
+            StatusMessage = statusMessage,
+            EntityType = TesTaskRunnerEntityType,
+            CorrelationId = correlationId ?? Guid.NewGuid().ToString(),
+            EntityId = entityId ?? Guid.NewGuid().ToString(),
+            Created = DateTime.UtcNow,
+            EventVersion = EventVersion,
+            EventDataVersion = EventDataVersion
+        };
+    }
+
+
+    private async Task PublishAsync(EventMessage message)
+    {
+        if (sinks.Count == 0)
+        {
+            logger.LogInformation("No sinks configured for publishing events");
+            return;
+        }
+
+        foreach (var sink in sinks)
+        {
+            logger.LogInformation($"Publishing event {message.Name} to sink: {sink.GetType().Name}");
+
+            await sink.PublishEventAsync(message);
+        }
+    }
+
+    public async Task FlushPublishersAsync(int waitTimeInSeconds = 60)
+    {
+        var stopTasks = sinks.Select(s => s.StopAsync());
+
+        await Task.WhenAll(stopTasks).WaitAsync(TimeSpan.FromSeconds(waitTimeInSeconds));
```

Comment:
I think TimeSpan would be better for waitTime, that way never confusion between "FromSeconds" and int

Reply:
No sure I see the benefit here, also I don't think it is possible to create a constant of type TimeSpan set to 60s.

Reply:
np, not a big deal, "TimeSpan is often preferred for readability and type safety when working with time-related operations. It makes the code self-documenting, and there's less chance of unit errors (e.g., mistakenly treating the delay as milliseconds rather than seconds)."

Reply:
i.e. `private static TimeSpan stopWaitDuration = TimeSpan.FromSeconds(60);`